apps/core/components/header/cart-badge.tsx

```diff
@@ -0,0 +1,37 @@
+'use client';
+
+import { Badge } from '@bigcommerce/components/badge';
+import { useEffect, useState } from 'react';
+import { z } from 'zod';
+
+export const CartQuantityResponseSchema = z.object({
+  count: z.number(),
+});
+
+interface CartBadgeProps {
+  count?: number;
+}
+
+export const CartBadge = ({ count }: CartBadgeProps) => {
+  const [fetchedCount, setFetchedCount] = useState<number | null>();
+  const computedCount = count ?? fetchedCount;
+
+  useEffect(() => {
+    async function fetchCartQuantity() {
+      const response = await fetch(`/api/cart-quantity/`);
+      const parsedData = CartQuantityResponseSchema.parse(await response.json());
+
+      setFetchedCount(parsedData.count);
+    }
+
+    if (count === undefined) {
+      void fetchCartQuantity();
+    }
+  }, [count]);
+
+  if (computedCount) {
+    return <Badge> {computedCount}</Badge>;
+  }
+
+  return null;
```

Comment:
I just thought of a better way after we chatted about this 😶‍🌫️ 
```suggestion
  const [fetchedCount, setFetchedCount] = useState<number | null>(count ?? null);

  useEffect(() => {
    async function fetchCartQuantity() {
      const response = await fetch(`/api/cart-quantity/`);
      const parsedData = CartQuantityResponseSchema.parse(await response.json());

      setFetchedCount(parsedData.count);
    }

    if (fetchedCount === null) {
      void fetchCartQuantity();
    }
  }, [fetchedCount]);

  if (!fetchedCount) {
    return null;
  }

  return <Badge> {fetchedCount}</Badge>;
```

Also, a :nit:, we try to `return null` before we return the component.